README.md

```diff
@@ -19,44 +19,241 @@ ICAT API to interface with the Data Gateway
+sqlite-devel openssl-devel xz xz-devel libffi-devel findutils
+```
+
+To make use of `pyenv`, let's install different versions of Python onto the system. In
+production, DataGateway API uses Python 3.6, so this should definitely be part a
+development environment for this repo. This stage might take some time as each Python
+version needs to be downloaded and built individually:
+
+```bash
+pyenv install 3.6.8
+pyenv install 3.7.7
+pyenv install 3.8.2
+```
+
+To verify the installation commands worked:
+
+```bash
+python3.6 --version
+python3.7 --version
+python3.8 --version
+```
+
+These Python versions need to be made available to local version of the repository. They
+will used during the Nox sessions, explained further down this file. Executing the
+following command will create a `.python-version` file inside the repo (this file is
+currently listed in `.gitignore`):
+
+```bash
+pyenv local 3.6.8 3.7.7 3.8.2
+```
+
+### Poetry (API Dependency Management)
+To maintain records of the API's dependencies,
+[Poetry](https://github.com/python-poetry/poetry) is used. To install, use the following
+command:
 
-## Setup and running the API   
-The database connection needs to be set up first. This is set in config.json, an example
-config file called `config.json.example` is provided.
+```bash
+curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -
+```
+
+The installation requires `~/.poetry/env` to be refreshed for changes to be applied.
+Open a new terminal or execute the following command to ensure the installation is
+completed smoothly:
+
+```bash
+source ~/.poetry/env
+```
+
+The dependencies for this repo are stored in `pyproject.toml`, with a more detailed
+version of this data in `poetry.lock`. The lock file is used to maintain the exact
+versions of dependencies from system to system. To install the dependencies, execute the
+following command (add `--no-dev` if you don't want the dev dependencies):
+
+```bash
+poetry install
+```
+
+To add a dependency to Poetry, run the following command (add `--dev` if it's a
+development related dependency). The
+[official docs](https://python-poetry.org/docs/cli/#add) give good detail regarding the
+intricacies of this command:
+
+```bash
+poetry add [PACKAGE-NAME]
+```
+
+### Nox (Automated Testing & Other Code Changes)
+When developing new features for the API, there are a number of Nox sessions that can be
+used to lint/format/test the code in the included `noxfile.py`. To install Nox, use Pip
+as shown below. Nox is not listed as a Poetry dependency because this has the potential
+to cause issues if Nox was executed inside Poetry (see
+[here](https://medium.com/@cjolowicz/nox-is-a-part-of-your-global-developer-environment-like-poetry-pre-commit-pyenv-or-pipx-1cdeba9198bd)
+for more detailed reasoning):
+
+```bash
+pip install --user --upgrade nox
+```
+
+To run the sessions defined in `nox.options.sessions` (see `noxfile.py`), simply run:
+
+```bash
+nox
```

Comment:
I got `PermissionError` when I tried running this command. It looks like the temporary files that are created with `tempfile.NamedTemporaryFile()` in line 11 and 53 in the `noxfile.py` file are stored in a directory that requires strict permissions. The workaround was to set the `dir` inside the `tempfile.NamedTemporaryFile()` call to a directory that does not require strict permissions (the project root directory in my case).

Reply:
This setting needs to be configurable either within this file or can be configured when executing any of the nox sessions (something like `nox -s lint -- --tmpdir /home/user1/`). I will test this on Linux, but this should also be tested on Windows.

Reply:
I've added this feature and documented it in https://github.com/ral-facilities/datagateway-api/pull/185/commits/0366d08d0b531c98d229e9ce3c849e00c2d8f043. Please could you test this and ensure it works? If you're happy with it, I'd say this is resolved.

Reply:
I tested this but it did not work despite trying with a few different directory paths, indicating that the `PermissionError` that I am seeing is not due to strict permissions being required for `nox` to create the temporary files. As a result, I spent some time investigating this and discovered that this is happening because Windows does not allow for processes other than the one that is used to create the temporary file to access it. This is documented in the library's [docs](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile):
`Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later)`. 

**Solution**:
The solution as per the comment in this [issue](https://github.com/bravoserver/bravo/issues/111#issuecomment-826990) is to set `NamedTemporaryFile`'s `delete` arg to `False` as it won't allow other processes to access the file otherwise. I am confused as to how this fixes the problem because looking at the function's docstring it says: `'delete' -- whether the file is deleted on close (default True)`. 

```Python
with tempfile.NamedTemporaryFile(delete=False) as requirements:
    session.run(... )
    session.install(...)
    requirements.close()
    os.unlink(requirements.name)
```

This needs to be tested on Linux and there is a chance that it may not work when the file is closed inside the `with` `as` block. If it does error, it might be better to not use the `with` `as` block and do it like this:
```Python
requirements = tempfile.NamedTemporaryFile(delete=False)
session.run(...)
session.install(...)
requirements.close()
os.unlink(requirements.name)
```

`.close()` and `.unlink()` can throw exceptions, so as shown in the example in the comment linked above, it's a good idea to put them inside a `try` `except` block.

Reply:
So I've added the `delete=False` and lines to manually close the file everywhere where `NamedTemporaryFile` is used. As a result, I've removed the `tmp_dir` option and all the code from that, just because it's not really needed now. I've tested this on Linux and happy it works, hopefully it works on Windows too.

Reply:
Looks good and works nicely. Could you please update the `README` as it contains info about the `tmp_dir` flag?

Reply:
As before, I've updated the README in the documentation branch, specifically commit https://github.com/ral-facilities/datagateway-api/commit/135965c02c1384d1350fe4052c759041b561a273. I've removed details regarding the flag and replaced it with a link to the the GitHub issue comment you found on solving the issue.